### ISSUE TITLE
[BUG] You current contribution: 0 samples

### ISSUE THREAD

**Issue by elricwan** [learning-at-home/hivemind#451]
**Describe the bug**
In the Albert example, when we set target_batch_size closer to the per_device_train_batch_size, then the peer could not contribute any samples. In my case, I set per_device_train_batch_size=4, target_batch_size=8, with one peer, then we get:

Step #109
Jan 19 18:01:52.012 [INFO] Your current contribution: 0 samples
Jan 19 18:01:52.012 [INFO] Performance: 33.865 samples/sec
Jan 19 18:01:52.012 [INFO] Local loss: 11.23390
Jan 19 18:01:52.033 [INFO] albert_experiment accumulated 0 samples for epoch #109 from 1 peers. ETA 0.21 sec (refresh in 0.50 sec)

If I increase the target_batch_size to 12, then the problem solved. Is there a rule about setting the per_device_train_batch_size and target_batch_size?

On the other hand, if I use two peers with target_batch_size=12 and per_device_train_batch_size=4, I also got Your current contribution: 0 samples, but albert_experiment accumulated 8 samples, so if I understand correctly, when we hace multiple peers, then the contribution would come from albert_experiment instead of one peer itself, right?

Jan 19 18:13:39.362 [INFO] Your current contribution: 0 samples
Jan 19 18:13:39.362 [INFO] Performance: 26.813 samples/sec
Jan 19 18:13:39.362 [INFO] Local loss: 10.77400
Jan 19 18:13:39.392 [INFO] albert_experiment accumulated 8 samples for epoch #182 from 2 peers. ETA 0.00 sec (refresh in 0.50 sec)

Apart from that, I have few questions:

1.  learning rate: if I understand it correctly, the choice of learning rate is scaled by target_batch_size, right?
2. Optimizer: there is two versions, one set use_local_updates=False, the other set use_local_updates=True. From my understanding, set use_local_updates=False is for large target_batch_size, and use_local_updates=True is like decentralized SGD which is good for small target_batch_size. Is there a detailed description about when to set use_local_updates=False, and when to set use_local_updates=True? Thanks.
3. Model save: with huggingface trainer, we usually set save_steps to save model. But with hivemind, we have CheckpointHandler and set store_checkpoints=True to save the model. I am wondering which one is better when using hivemind? If we use save_steps to save the model, we don't need to define CheckpointHandler in run_training_monitor.py file, right?

**To Reproduce**
For monitor, I run:

python run_training_monitor.py \
--experiment_prefix albert_experiment \
--wandb_project albert_wandb

For trainer, I run:

IP=/ip4/192.168.0.188/tcp/40587/p2p/QmTgM1kgdmdyxLgnCzNU6s8jHCcpvdCjZBHiCkPtLMhMbF

WANDB_DISABLED=true CUDA_VISIBLE_DEVICES=0 python run_trainer.py \
--experiment_prefix albert_experiment  \
--initial_peers $IP  \
--logging_first_step  \
--output_dir ./outputs \
--overwrite_output_dir \
--logging_dir ./logs \
--dataset_path="/home/protago/Xiangpeng/hivemind/examples/albert/data/albert_tokenized_wikitext" \
--per_device_train_batch_size 4 \
--warmup_ratio 0 \
--learning_rate 0.00176 \
--num_train_epochs=5 \
--target_batch_size=12 \
--save_steps=1000000

IP=/ip4/192.168.0.188/tcp/40587/p2p/QmTgM1kgdmdyxLgnCzNU6s8jHCcpvdCjZBHiCkPtLMhMbF

WANDB_DISABLED=true CUDA_VISIBLE_DEVICES=1 python run_trainer.py \
--experiment_prefix albert_experiment  \
--initial_peers $IP  \
--logging_first_step  \
--output_dir ./outputs \
--overwrite_output_dir \
--logging_dir ./logs \
--dataset_path="/home/protago/Xiangpeng/hivemind/examples/albert/data/albert_tokenized_wikitext" \
--per_device_train_batch_size 4 \
--warmup_ratio 0 \
--learning_rate 0.00176 \
--num_train_epochs=5 \
--target_batch_size=12 \
--save_steps=1000000

**Environment**
Please list:
* python version (e.g. 3.8.1); 3.8
* hivemind.__version__; 1.1.0
* Please copy and paste the output from pytorch [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)

If the script doesn't work, please report pytorch and numpy versions manually. We also encourage you to include any additional information that you believe can help us solve the issue.

**Comment by justheuristic** [comment]
Hi!

A good initial guide to setting target batch size can be found in the Optimizer docs:
https://learning-at-home.readthedocs.io/en/latest/modules/optim.html#hivemind.optim.optimizer.Optimizer

A good rule of thumb is that __target batch size must be large enough so that all peers need to make 10+ local iterations to achieve it__. So, if you have 50 peers with batch size 4 each, then 50peers x 10 iterations x 4samples/iteration = 2000. So a batch size of 2048 would suffice. Another consideration is that peers should be able to aggregate their gradients faster than they compute the next batch. So, if the network is very slow on some peers, you may want to further increase batch size, e.g. to 4096 or even more.

In turn, the __per_device_train_batch_size__ is how many samples peer runs in one go. It is entirely dependent on peer's GPU. As a good rule of thumb, pick the largest batch size that fits into your GPU as long as it takes less than 10 seconds to compute forward+backward on that batch. The reason behind this is that larger batches generally lead to higher training throughput per GPU.

In your specific setup, this is actually the intended behavior for an edge. Your `target_batch_size` is so small that even a single peer can get the full batch size in one iteration. So whichever peer reports the batch first actually gets all the contribution while the other one just tags along.

**Comment by justheuristic** [comment]
> learning rate: if I understand it correctly, the choice of learning rate is scaled by target_batch_size, right?

We use LAMB optimizer that provides recommended learning rate scaling rules. Please see table 4 in https://arxiv.org/abs/1904.00962 . They also provide good intuition on how to adapt that rule to other models


> Optimizer: there is two versions, one set use_local_updates=False, the other set use_local_updates=True. From my understanding, set use_local_updates=False is for large target_batch_size, and use_local_updates=True is like decentralized SGD which is good for small target_batch_size. Is there a detailed description about when to set use_local_updates=False, and when to set use_local_updates=True? Thanks.

You got the idea right, but there's an important technical detail: you need to set large target_batch_size in both cases

- if use_local_updates = False, you accumulate this batch size and perform one synchronous gradient update
- if use_local_updates = True, peers run optimizer steps __on every local iteration__, e.g. every 4 samples if that's your per_device_train_batch_size

In case when use_local_updates = True, target_batch_size is used for auxiliary purposes: (1) peers average parameters in background after accumulating target_batch_size samples and (2) learning rate scheduler is tied to the number of times peers accumulated target_batch_size. More information and the rationale behind this can be found [in the optimizer docs](https://learning-at-home.readthedocs.io/en/latest/modules/optim.html)


>  Model save: with huggingface trainer, we usually set save_steps to save model. But with hivemind, we have CheckpointHandler and set store_checkpoints=True to save the model. I am wondering which one is better when using hivemind? If we use save_steps to save the model, we don't need to define CheckpointHandler in run_training_monitor.py file, right?

Both would work, and the choice depends on your application. Here's a few examples:

- when we trained bert models with Neuropark and NCAI communities, most of the participants were volunteers that only contribute part-time. For them, it is (aguably) easier to download the model from huggingface after the training has concluded. Hence, we had an auxiliary peer (monitor) that was charged with saving model and uploading it to HF hub every few hours.

- in my internal experiments with spot instances, i'm too too lazy to even start a monitor: i just connect GPU peers to other GPU peers as `--initial_peers` and call it a day. In that case it is indeed easier to save the model from normal peers.

**Comment by elricwan** [comment]
Thank you for the instruction! Now I have a better understanding about how to use hivemind framework, big help!
